homeassistant/components/icloud/account.py

```diff
@@ -110,9 +115,36 @@ def setup(self) -> None:
                 self._icloud_dir.path,
                 with_family=self._with_family,
             )
-        except PyiCloudFailedLoginException as error:
+        except PyiCloudFailedLoginException:
             self.api = None
-            _LOGGER.error("Error logging into iCloud Service: %s", error)
+            # Login failed which means credentials need to be updated.
+            username = self._config_entry.data[CONF_USERNAME]
+            message = (
+                f"Your password for '{username}' is no longer working. Go to the "
```

Comment:
I did see that issue and didn't realize it was related to that issue at the time, but I was getting a bunch of authentication alerts on my phone, which led me to change my iCloud password, which I then needed to update in HomeAssistant, which is what led to this PR.

Regarding your idea, we should certainly get a fresh token during reauth, but my understanding is that `PyiCloudFailedLoginException` only fires if your credentials are incorrect as seen here: https://github.com/picklepete/pyicloud/blob/4075c41ecc07f03e1994ae1d0996907d767bd2d3/pyicloud/base.py#L236

So we can try classic auth first before asking for a password change, but I don't know if it will resolve the issue.

Reply:
I just checked and the code should get a fresh token during auth because we recreate `PyiCloudService` which runs the `authenticate` step.

Reply:
I am also not clear on what "trying classic auth first before asking for a password change" entails since the trigger for the reauth is a failure of the authentication step

Reply:
Will check the auth token issue when I have time, and propose a fix in a new PR.

Or maybe you can propose a fix too ?

Reply:
Do you want me to pursue the suggestion I had made? I can't figure out where to detect when the token has expired, hence the schedule idea, but if you have an idea I would be happy to put a PR up for the change